integration/runner/yugabyte.go

```diff
@@ -47,47 +67,187 @@ func StartYugaCluster(ctx context.Context, t *testing.T, clusterSize uint) (
-	for range clusterSize {
-		cluster.AddNode(ctx, t)
+	cluster := &YugaClusterController{
+		networkName: uuid.NewString(),
 	}
+	dbtest.CreateDockerNetwork(t, cluster.networkName)
+	t.Cleanup(func() {
+		dbtest.RemoveDockerNetwork(t, cluster.networkName)
+	})
+
+	// We create the nodes before startup to ensure
+	// we have the names of the master nodes.
+	// This information is required ahead of time to start each node.
+	for range numberOfMasters {
+		cluster.createNode(MasterNode)
+	}
+	for range numberOfTablets {
+		cluster.createNode(TabletNode)
+	}
+	cluster.startNodes(ctx, t)
 
 	t.Cleanup(func() {
 		cluster.stopAndRemoveCluster(t)
 	})
 
-	clusterConnection := cluster.getNodesConnections(ctx, t)
+	// The master nodes are not involved in DB communication;
+	// the application connects to the tablet servers.
+	clusterConnection := cluster.getConnectionsOfGivenRole(ctx, t, TabletNode)
 	clusterConnection.LoadBalance = true
 
 	return cluster, clusterConnection
 }
 
-// AddNode creates, starts and add a YugabyteDB node to the cluster.
-func (cc *YugaClusterController) AddNode(ctx context.Context, t *testing.T) *dbtest.DatabaseContainer {
-	t.Helper()
-
+func (cc *YugaClusterController) createNode(role string) {
 	node := &dbtest.DatabaseContainer{
-		Name:         fmt.Sprintf("yuga-%s", uuid.New()),
-		Role:         LeaderNode,
+		Name:         fmt.Sprintf("yuga-%s-%s", role, uuid.New().String()),
+		Image:        defaultImage,
+		Role:         role,
 		DatabaseType: dbtest.YugaDBType,
-		Cmd:          dbtest.YugabyteCMD,
+		Network:      cc.networkName,
 	}
+	cc.nodes = append(cc.nodes, node)
+}
 
-	if cc.GetClusterSize() != 0 {
-		// node.Role refers to the node's Raft role.
-		node.Role = FollowerNode
-		node.Cmd = append(node.Cmd, fmt.Sprintf("--join=%s", cc.getLeaderHost(ctx, t)))
+func (cc *YugaClusterController) startNodes(ctx context.Context, t *testing.T) {
+	t.Helper()
+	for _, n := range cc.IterNodesByRole(MasterNode) {
+		n.Cmd = nodeConfig(n.Role, n.Name, cc.getMasterAddresses(), cc.desiredRF())
+		n.StartContainer(ctx, t)
 	}
 
-	cc.nodes = append(cc.nodes, node)
+	expectedAlive := len(maps.Collect(cc.IterNodesByRole(MasterNode)))
+	require.EventuallyWithT(t, func(ct *assert.CollectT) {
+		_, actualAlive := cc.getLeaderMaster(t)
+		require.Equal(ct, expectedAlive, actualAlive)
+	}, time.Minute, time.Millisecond*100)
+
+	for _, n := range cc.IterNodesByRole(TabletNode) {
+		n.Cmd = nodeConfig(n.Role, n.Name, cc.getMasterAddresses(), cc.desiredRF())
+		n.StartContainer(ctx, t)
+		n.EnsureNodeReadiness(t, "syncing data to disk ... ok")
+	}
+}
 
-	require.NoError(t, nodeStartupRetry.Execute(ctx, func() error {
-		t.Logf("starting db node %v with role: %v", node.Name, node.Role)
-		node.StartContainer(ctx, t)
-		return node.EnsureNodeReadiness(t, "Data placement constraint successfully verified")
-	}))
+func (cc *YugaClusterController) getMasterAddresses() string {
+	var masterAddresses []string //nolint:prealloc
```

Comment:
minor: please justify any "nolint" as a comment.

```suggestion
	var masterAddresses []string //nolint:prealloc // <justification>.
```